joomla/joomla-cms
# [4.0][RFC] Google fonts localiser for all

Pull Request for Issue https://github.com/joomla/cassiopeia/issues/129 .

### Some facts on Google Fonts
- They probably exist in the user's cache so they're super fast
This is utterly wrong for 2 reasons: A. Assumptions are really bad when it comes to render blocking elements B. Since the revealing of Spectre and Meltdown vulnerabilities **all** browsers run every site in isolation, so there are no global cache that the fonts could benefit from.

- CDNs are fast
Yes and no. If you observe their own performance in isolation yes their fast but if you're adding them in the loop of a website you have to account the extra DNS resolving, the extra connection, etc

- You can mitigate this with preconnect and prefetch
To a degree but not completely. Unfortunately there are 2 different domains, one for the css file and another for the actual fonts. So even with prefetch and preconnect the problem is not resolved (decreasing the time, but not eliminating it)


- We can tolerate this performance degradation in favour of DX
Actually this PR is meant to be a proof that you can have your cake and eat it. Keep the DX high with a nice simple widget that will localise the google fonts, so also the end users don't have to pay the high tax of blank screen

### Performance audit
For those not in the loop of auditing sites for performance, here's a simple explanation of what we try to mitigate here:
Goto `https://www.webpagetest.org/easy.php` enter `joomla.org` and click on the 2 checkboxes. Await for the results or check them here: https://www.webpagetest.org/result/201003_Di3P_561f21bbe231da8eea0547d2be0de7c9/2/details/#waterfall_view_step1

The CSS CDN for the google font takes over 1 sec to d/l the 0.5kb of the data. (that file needs to be parsed as well)
![Screenshot 2020-10-04 at 01 23 19](https://user-images.githubusercontent.com/3889375/95013071-f64dd600-063d-11eb-884b-e1eff438d0a4.png)

The CDN for the actual fonts takes around 1.5sec to d/l the font
![Screenshot 2020-10-04 at 01 23 51](https://user-images.githubusercontent.com/3889375/95013096-30b77300-063e-11eb-83e1-31e101b1a64b.png)

In total you have a 2.5 sec delay (on a real 3g connection) on top of the other render blocking stuff (namely Bootstrap and all the javascript files that AREN'T deferred)

Eliminating this delay **will make EVERY Joomla site faster**

### Summary of Changes
The proposal basically comes in 2 parts:
- A new Form Field with a very basic (atm) client side app (it should support setting: the font name, the sizes, italics, etc but right now it just accepts the name)

- A PHP endpoint to store the font/css. Mind that since basically we're talking about uploading files to the server this is inherently unsafe.

I took some time to do some extra checks, namely: 
- The field will echo a timestamp which is saved in the server's state, so random XHR's won't work without form token, user privileges AND the extra timestamp.
- The files are mime checked, for css the check will bail on php files and for font files if it doesn't meet the expected file mime 
- Client side (it's not implemented) the usage of private fields will ensure that the flow of data can not be hacked

### Testing Instructions
You need to run `rollup -c .rollup.config.js` on the root folder to compile the js

Then goto the Atum styles and check the advanced tab you should see a new field at the end:
![Screenshot 2020-10-04 at 12 50 09](https://user-images.githubusercontent.com/3889375/95013385-2ac29180-0640-11eb-9748-b8df9a150f56.png)

Insert any Google Font name and the click on the localise it button. Check the Atum folder:
![Screenshot 2020-10-04 at 12 51 40](https://user-images.githubusercontent.com/3889375/95013406-5d6c8a00-0640-11eb-95f2-5a847be420a2.png)

That's all


@richard67
